### PR TITLE
More Typed AST

### PR DESCRIPTION
**Pull request by saem**

### CODE REVIEW

File: cps/environment.nim [hunk 1]
```diff
@@ -65,35 +66,36 @@ func isDirty*(e: Env): bool =
...
 
-proc castToRoot(e: Env; n: NimNode): NimNode =
-  newTree(nnkCall, e.root, n)
+proc castToRoot(e: Env; n: NormalizedNode): NormalizedNode =
+  newCall(e.root, n)
 
-proc castToChild(e: Env; n: NimNode): NimNode =
-  newTree(nnkCall, e.identity, n)
+proc castToChild(e: Env; n: Name): NormalizedNode =
+  # XXX: remove NimNode
```
Comment (alaviss) [inline cps/environment.nim:79]: What `NimNode`?

File: cps/normalizedast.nim [hunk 2]
```diff
@@ -1,66 +1,452 @@
...
+  r.Name
+
+proc desym*(n: Name): Name {.borrow.}
+  ## ensures that `Name` is an `nnkIdent`
+proc resym*(fragment: NormalizedNode, sym, replacement: Name): NormalizedNode {.borrow.}
+  ## replace `sym` in the AST `fragment` with the `replacement`
+
+func typeInst*(n: Name): NimNode =
+  ## gets the type via `getTypeInst`
+  ## XXX: this shouldn't be done on name, as these could be bare idents
```
Comment (alaviss) [inline cps/normalizedast.nim:374]: Then I think this should either not exist, or add a `deprecated` warning to it so we can refactor the parts that use it

File: cps/normalizedast.nim [hunk 3]
```diff
@@ -72,39 +458,61 @@ func hasType*(n: DefLike): bool =
   ## has a non-Empty type (`typ`) defined
   n.typ.kind != nnkEmpty
 
-func inferTypFromImpl*(n: DefLike): NimNode =
+func inferTypFromImpl*(n: DefLike): Name =
```
Comment (alaviss) [inline cps/normalizedast.nim:461]: `TypeExpr`?

File: cps/normalizedast.nim [hunk 4]
```diff
@@ -113,6 +521,23 @@ func hasType(n: VarLetLike): bool = n.def.typ.kind != nnkEmpty
...
+    ## abstract over let or var sections with identdefs within
+
+func identdef*(n: VarLetIdentDefLike): IdentDef = n.def.IdentDef
+  ## retrieve the innner IdentDef
+func name*(n: VarLetIdentDefLike): Name = n.identdef.name
+  ## Name (ident|sym) of the identifer, as we only have a single identdefs it
+  ## will have the name
+func inferTypFromImpl*(n: VarLetIdentDefLike): Name =
+  ## returns the typ if specified or uses `macro.getTypeImpl` to infer it
+  if n.hasType: n.typ.Name else: getTypeImpl(n.val).Name
```
Comment (alaviss) [inline cps/normalizedast.nim:537]: This should be removed. The promise we make in `normalizingRewrites` is that there will always be a type.

File: cps/normalizedast.nim [hunk 5]
```diff
@@ -1,66 +1,452 @@
...
+    nnkVarTy, nnkRefTy, # the weirder ones
+    nnkTupleConstr      # still have tuples in a few areas
+    # nnkEmpty          # excluded as it's only allowed in some cases
+  }
+  ## list of type NimNodeKind
+  ## XXX: this is an incomplete list
 
 func errorGot(msg: string, n: NimNode, got: string = repr(n)) =
   ## useful for error messages
   error msg & ", got:\n" & repr(got), n
```
Comment (alaviss) [inline cps/normalizedast.nim:146]: If this is only used for validation in here, make it raise `ValueError` instead.

File: cps/normalizedast.nim [hunk 6]
```diff
@@ -188,69 +600,146 @@ func clone*(n: VarLet, value: NimNode = nil): VarLet =
...
-converter identDefVarToIdentDefVarLet*(n: IdentDefVar): IdentDefVarLet =
-  # allow downgrading
-  n.IdentDefVarLet
+proc newProcDef*(name: Name, formalParams: openArray[NimNode]): ProcDef =
+  ## create a new proc def with name and formal params (includes return type)
+  ## and an empty body (`nnkStmtList`)
+  newProc(name.Nimnode, formalParams, newStmtList()).ProcDef
 
-# fn-ProcDef
+func asProcDef*(n: NimNode): ProcDef =
```
Comment (alaviss) [inline cps/normalizedast.nim:674]:
```suggestion
func asProcDef*(n: NormalizedNode): ProcDef =
```

File: cps/rewrites.nim [hunk 7]
```diff
@@ -397,8 +434,24 @@ func replace*(n: NimNode, match: Matcher, replacement: NimNode): NimNode =
       copyNimTree replacement
     else:
       nil
-
+
```
Comment (alaviss) [inline cps/rewrites.nim:437]: trailing spaces

**Review by alaviss — COMMENTED** [review]
Looks much better that what we have right now.

Can you move `genField` to `normalizedast`?

File: cps/normalizedast.nim (hunk 5 above)
Comment (saem) [reply cps/normalizedast.nim:146]: That's not great for debugging is it?

Comment (saem) [reply cps/normalizedast.nim:146]: Is that true for proc params and ident defs inside a var or let section?

The other issue is that they dervive type information differently.

File: cps/normalizedast.nim (hunk 6 above)
Comment (saem) [reply cps/normalizedast.nim:674]: That means a bunch of changes to other code in order for us to have a `NormalizedNode` in the first place.

Most annoying part is in `transform.cpsResolver`

File: cps/normalizedast.nim (hunk 4 above)
Comment (saem) [reply cps/normalizedast.nim:537]: Is that true for proc params and ident defs inside a var or let section?

The other issue is that they dervive type information differently.

**Comment by disruptek** [comment]
Do we just merge it so we can move forward with this enterprise?

**Comment by saem** [comment]
> Do we just merge it so we can move forward with this enterprise?

I'll look at it again this weekend, still feels like it should do more with less.

**Pull request merged**